borsh/src/schema.rs

```diff
@@ -299,6 +301,52 @@ where
     }
 }
 
+impl<K, V> BorshSchema for BTreeMap<K, V>
+where
+    K: BorshSchema,
+    V: BorshSchema,
+{
+    fn add_definitions_recursively(definitions: &mut HashMap<Declaration, Definition>) {
+        let definition = Definition::Sequence {
+            elements: <(K, V)>::declaration(),
+        };
+        Self::add_definition(Self::declaration(), definition, definitions);
+        <(K, V)>::add_definitions_recursively(definitions);
+    }
+
+    fn declaration() -> Declaration {
+        format!(r#"BTreeMap<{}, {}>"#, K::declaration(), V::declaration())
+    }
+}
+
+impl<T> BorshSchema for BTreeSet<T>
+where
+    T: BorshSchema,
+{
+    fn add_definitions_recursively(definitions: &mut HashMap<Declaration, Definition>) {
+        let definition = Definition::Sequence {
+            elements: <T>::declaration(),
+        };
+        Self::add_definition(Self::declaration(), definition, definitions);
+        <T>::add_definitions_recursively(definitions);
+    }
+
+    fn declaration() -> Declaration {
+        format!(r#"BTreeSet<{}>"#, T::declaration())
+    }
+}
+
+// Because it's a zero-sized marker, its type parameter doesn't need to be
+// included in the schema and so it's not bound to `BorshSchema`
+impl<T> BorshSchema for PhantomData<T> {
```

Comment:
On the fence about this one, love to hear other's thoughts.

Seems like a field like this should be ignored from the schema rather than inserting this. Not sure if implementing this would cause issues with maintenance if we ever wanted to remove.

Also, wonder if it could be `nil` as `()` is

Reply:
I agree with @austinabell, I don't see any value in exposing Rust type-system artifacts.

Reply:
> This is a valid point, I mean I assume that perhaps BorshSchema should be 1:1 with borsh impls, which does contain PhantomData, but I'm not sure if that is a guarantee we want/have with schema?

Yeah, I think that's a reasonable thing to do

> As for using nil vs having a specific case for PhantomData, I'm not sure this adds value in any way. Given that for borsh (and even serde) this is just (de)serialized as nil, when would someone care if something was a PhantomData or not?

True, I don't think it does add any value and if anything it really should be hidden from schema because it is a language specific construct. Let's change this to "nil" then.

Reply:
@austinabell I have just noticed that we dropped the conversation here. I am not sure about "nil", though, is it equivalent to skip? Can we actually just skip these fields altogether?

Reply:
yeah, nil in borsh will not write any bytes. I think the current state of this PR is fine. Do you have a reason for why this declaration would not match the borsh impl?

Reply:
> Or are you suggesting that we remove borsh impl from PhantomData for symmetry and forcing devs to skip it?

What I wanted to say is that I don't see a need for PhantomData to be surfaced in the schema at all, so treating it as skipped field rather than nil-type field. I am thinking from the perspective of handling the schema in other languages besides Rust will look odd to have this nil occasionally for no good reason.

@austinabell I wanted to only inform you about the way of thinking I have, but I don't have strong opinion about BorshSchema and I am also not even a contributor there, so feel free to cast your vote and merge if you believe this is good to go 👍

Reply:
Borsh schema exposes field names and with nil type you will still see the unnecessary field name in the schema. However, thinking about it now, it might be confusing if we skip some fields out of blue even if they are virtual, so maybe we should just recommend using borsh(skip) explicitly and then I would consider NOT implementing BorshSchema for PhantomData

Reply:
@itegulov  this issue is about schema, not serialization/deserialization. Borsh ser/de just like serde implements traits for PhantomData, however, BorshSchema was left not implemented for PhantomData and now I argue that it should not be, thus forcing users explicitly skip the field from the schema

Reply:
@frol Oh yeah right I got confused for a second, ignore the second part of my comment!

Reply:
I don't see why there would be asymmetry there. If it's implemented for Borsh, why not include it in the schema? By the same logic, should we not remove `impl BorshSchema for ()`?